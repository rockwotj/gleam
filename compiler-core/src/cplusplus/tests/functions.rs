use crate::assert_cpp;

#[test]
fn exported_functions() {
    assert_cpp!(
        r#"
pub fn add(x, y) {
    x + y
}"#,
    );
}

#[test]
fn private_functions() {
    assert_cpp!(
        r#"
fn add(x, y) {
    x + y
}"#,
    );
}

#[test]
fn anonymous_functions() {
    assert_cpp!(
        r#"
fn squared(x) {
  let y = fn(z) { z * z }
  y(x)
}
    "#
    )
}

#[test]
fn captured_variables() {
    assert_cpp!(
        r#"
fn do_stuff(x, y) {
  let f = fn(z) { { x * y } + z }
  f(x + y)
}
    "#
    )
}

#[test]
fn multiple_module_functions() {
    assert_cpp!(
        r#"
fn add(x, y) {
  x + y
}
fn subtract(x, y) {
  add(x, 0 - y)
}
    "#
    )
}

#[test]
fn generic_function() {
    assert_cpp!(
        r#"
fn identity(x) {
  x
}
fn wrapped(x) {
  identity(x)
}
        "#
        )
}

#[test]
fn specialized() {
    assert_cpp!(
        r#"
fn identity(x) {
  x
}
fn add(a, b) {
  identity(a) + identity(b)
}
        "#
        )
}
