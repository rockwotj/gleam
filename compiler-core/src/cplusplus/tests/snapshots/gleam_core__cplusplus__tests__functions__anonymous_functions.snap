---
source: compiler-core/src/cplusplus/tests/functions.rs
expression: "\nfn squared(x) {\n  let y = fn(z) { z * z }\n  y(x)\n}\n    "
---
#ifndef MY_PACKAGE_MY_MODULE_H_
#define MY_PACKAGE_MY_MODULE_H_

#include <gleam.h>

namespace my {
namespace module {

int64_t squared(int64_t x);

} // namespace my
} // namespace module

#endif // MY_PACKAGE_MY_MODULE_H_

---
#include "module.h"

namespace my {
namespace module {

int64_t squared(int64_t x) {
  gleam::Function<int64_t, int64_t> y = [=](int64_t z) -> int64_t {
    return z * z;
  };
  return y(x);
};

} // namespace my
} // namespace module

